src/array/cuda/atomic.cuh

```diff
@@ -228,23 +238,33 @@ __device__ __forceinline__ double AtomicAdd<double>(double* addr, double val) {
 #if __CUDA_ARCH__ >= 600
   return atomicAdd(addr, val);
 #else
-  return *addr + val;
+  typedef double T;
+  typedef typename Cast<T>::Type CT;
+  CT* addr_as_ui = reinterpret_cast<CT*>(addr);
+  CT old = *addr_as_ui;
+  CT assumed = old;
+  do {
+    assumed = old;
+    old = atomicCAS(addr_as_ui, assumed,
+        Cast<T>::Encode(Cast<T>::Decode(old) + val));
+  } while (assumed != old);
+  return Cast<T>::Decode(old);
 #endif
 }
 
 #ifdef USE_FP16
 #if defined(CUDART_VERSION) && CUDART_VERSION >= 10000
-// half make sure we have half support
-#if __CUDA_ARCH__ >= 600
 template <>
 __device__ __forceinline__ half AtomicAdd<half>(half* addr, half val) {
+// half make sure we have half support
 #if __CUDA_ARCH__ >= 700
   return atomicAdd(addr, val);
 #else
-  return *addr + val;
+  half res = half(float(*addr) + float(val));
+  *addr = res;
+  return res;
```

Comment:
I think atomic functions should return the value stored in `addr`. For example in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicadd, 
> reads the 16-bit, 32-bit or 64-bit word `old` located at the address `address` in global or shared memory, computes `(old + val)`, and stores the result back to memory at the same address. These three operations are performed in one atomic transaction. The function returns `old`.

Reply:
Good catch!  Thanks.  I've fixed it now.